sys/net/application_layer/gcoap/forward_proxy.c

```diff
@@ -424,28 +452,31 @@ static int _gcoap_forward_proxy_via_coap(coap_pkt_t *client_pkt,
 
     unsigned token_len = coap_get_token_len(client_pkt);
 
-    coap_pkt_init(&pkt, proxy_req_buf, CONFIG_GCOAP_PDU_BUF_SIZE,
+    coap_pkt_init(&client_ep->pdu, proxy_req_buf, CONFIG_GCOAP_PDU_BUF_SIZE,
                   sizeof(coap_hdr_t) + token_len);
 
-    pkt.hdr->ver_t_tkl = client_pkt->hdr->ver_t_tkl;
-    pkt.hdr->code = client_pkt->hdr->code;
-    pkt.hdr->id = client_pkt->hdr->id;
+    client_ep->pdu.hdr->ver_t_tkl = client_pkt->hdr->ver_t_tkl;
+    client_ep->pdu.hdr->code = client_pkt->hdr->code;
+    client_ep->pdu.hdr->id = client_pkt->hdr->id;
 
     if (token_len) {
-        memcpy(coap_get_token(&pkt), coap_get_token(client_pkt), token_len);
+        memcpy(coap_get_token(&client_ep->pdu), coap_get_token(client_pkt), token_len);
     }
 
     /* copy all options from client_pkt to pkt */
-    len = _gcoap_forward_proxy_copy_options(&pkt, client_pkt, client_ep, urip);
+    len = _gcoap_forward_proxy_copy_options(&client_ep->pdu, client_pkt, client_ep, urip);
 
     if (len == -EINVAL) {
         return -EINVAL;
     }
 
-    len = gcoap_req_send((uint8_t *)pkt.hdr, len,
-                         &origin_server_ep,
-                         _forward_resp_handler, (void *)client_ep,
-                         GCOAP_SOCKET_TYPE_UNDEF);
+    /* DTLS communication is blocking the gcoap thread,
+     * therefore the communication should be handled in the proxy thread */
+
+    msg_t msg = {   .type = FORWARD_PROXY_MSG_SEND,
+                    .content.ptr = client_ep };
+    msg_try_send(&msg, _forward_proxy_pid);
```

Comment:
You can then use
```suggestion
    if (IS_USED(MODULE_GCOAP_FORWARD_PROXY_THREAD)) {
      /* WORKAROUND: DTLS communication is blocking the gcoap thread,
       * therefore the communication should be handled in the proxy thread */
  
         msg_t msg = {   .type = FORWARD_PROXY_MSG_SEND,
                         .content.ptr = client_ep };
         msg_try_send(&msg, _forward_proxy_pid);
    }
    else {
         len = gcoap_req_send((uint8_t *)pkt.hdr, len,
                              &origin_server_ep,
                              _forward_resp_handler, (void *)client_ep,
                              GCOAP_SOCKET_TYPE_UNDEF);
    }
```

What worries me a bit is that this changes the semantic of `len`, but this is a workaround after all and is already the case in your current code.